### DISCUSSION TITLE
How can I use both getStaticProps (static page) and normal dynamic page if user is logged in?

### DISCUSSION THREAD

**Discussion by rahul3103** [vercel/next.js#17248] [Help] [answered]
I want to have a static page for a route when user is not logged in but when he logs in, then on the same route I have to show dynamic content according to user. Is it possible to do that on next js? Like completely two different sets of page.

**Comment by leerob** [comment, marked as answer]
Yes! Here's an example:

1. The page is automatically statically generated since there's no blocking data
2. The `useUser()` hooks allows us to fetch the user on the client-side

https://github.com/vvo/next-iron-session/blob/master/examples/next.js/pages/profile-sg.jsx

**Reply by leerob** [reply]
Let me know if you have any other questions 😄

**Reply by tjdam** [reply]
Hi @leerob!
How is profile-sg.jsx statically generated without getStaticProps + getStaticPaths?

**Reply by leerob** [reply]
Give these a read:

- https://nextjs.org/docs/basic-features/pages#static-generation-without-data
- https://nextjs.org/docs/advanced-features/automatic-static-optimization

**Reply by tjdam** [reply]
Ohhh this is really exciting!

One question I have is... what if my page does depend on data and I'm using getStaticProps but the data differs if the user is logged in and if it's not logged in?

Is there a canonical way to bypass SSG and use a hook, for instance?
Or maybe there is another strategy to achieve that?

**Reply by leerob** [reply]
If you're trying to conditionally render data based on whether a user is logged in, you can:

- Use SSR with `getServerSideProps`
- Use CSR by showing a loading state/skeleton and then fetching the user. If there's a user, show different content.

**Reply by tjdam** [reply]
So it should be possible for this state/skeleton you mention to be SSG and then have it CSR on top, right?

**Reply by leerob** [reply]
Correct, that is possible. For example, you could fetch a bunch of data at build time with SSG and then use SWR to continue fetching new data client-side.

**Reply by lucasbasquerotto** [reply]
> * Use SSR with `getServerSideProps`

@leerob If I understand correctly:

> If a page uses Server-side Rendering, the page HTML is generated on each request.

Coudn't I use SSG with `getStaticProps` on these 2 possible scenarios:

1) **User isn't logged**: Load the page content in the server, **serve as a static page only for unlogged users**, revalidate after 5 minutes.

2) **Logged user** (**same path**): Return a spinner / skeleton as a **static page only for logged users** and **load the data on client side**, use the same static page even if the id of the page differs (like a post id, because the skeleton will be the same, although it would still be ok if a different static page is generated for different ids), revalidate after 1 day (or more).

**Is the above achievable with NextJS?**

**Reply by leerob** [reply]
Yes, you're free to use SSR or CSR on a per-page basis 👍  Whichever rendering method works better for you.

**Reply by lucasbasquerotto** [reply]
@leerob Thanks for the reply, but in my case I would like to use **SSG** for the same page (same url) but the static page returned would be different based on the fact of the user being logged in or not. Is it achievable? 

Another possibility would be to put a logged in slug in the path, like `/:logged/mypage`, so that `/true/mypage` and `/false/mypage` would return different static pages. I think this is already possible to do, but it would be necessary to do that **without changing the url in the browser** (all users would see the same url, but internally NextJS would return a different page based on the fact that the user is logged in or not). Is it possible without a custom server as a proxy in the middle?

**Reply by leerob** [reply]
You should use CSR for that (which could be in combination with SSG for the initial loading state).

**Reply by lucasbasquerotto** [reply]
Yes, I would use CSR to fetch data at the client side **when the user is logged in** (the server would return a static page with only a spinner / skeleton). But **when the user is not logged** (or is a crawler or search engine), I would like to serve a static page that has the data already fetched (so it would be 2 different cases, 1 for the case that the user is logged, another in the case the user is not logged, both would return different static pages). It seems I can do that if I use different urls (like the example I gave before, `/:logged/mypage`), but can I do that using the same url?

**Reply by pitasi** [reply]
@lucasbasquerotto I think that's exactly what he's doing in the example he linked before: https://github.com/vvo/next-iron-session/blob/master/examples/next.js/pages/profile-sg.jsx.

See `return <Layout>loading...</Layout>;`, that would be your "not logged in" page. Then, on client side you can start the spinner while you requests are loading.

**Reply by lucasbasquerotto** [reply]
@Pitasi I don't think so. If I understood the example correctly, it will return the same static page from the server no matter if the user is logged or not, the page having a loading message, fetching the data if the user is logged, or redirecting to a login page if the user isn't logged, this being done at the **client** side.

What I want is to fetch the data at the server and return a **static** page **with the data already loaded** if the user is **not** logged, especially for search engines. If the user is logged and access the **same** url, it should return a static page with a spinner / skeleton and load the data at the client side.

I think the above approach should be very common (or maybe not because incremental SSG is relatively new), because you can have a better SEO and at the same time it should reduce a lot processing at the server side (because there's no SSR). If the user is logged it would still fetch the data needed at the client side, but this is expected anyway, even if using CRA.

I think I can achieve what I want proxying to different urls from the nginx server, based on the existence of the authentication cookie, but I would prefer a solution in the app. I will try and see if it works (with a proxy), although I don't know if it will because when the page returns from the server, the client will think it's a different page because the url (and the page component) used in the server was different from the one used in the browser (and may try to load again), but if I use something like redux I may be able to avoid loading the data again (using the data already loaded from the store), although the page itself might still be loaded.

**Reply by amaslakov** [reply]
@lucasbasquerotto have you had any success serving SSG page and having CSR on the same page when the user is logged in, using the same url?

**Comment by rahul3103** [comment]
@leerob So, the static page which I want is completely different from logged in page. It's two different containers all together. So, we thought if user is not logged in we gonna make a static page and other wise dynamic content for the logged in user which uses completely different container all together

**Reply by leerob** [reply]
See answer here: https://github.com/vercel/next.js/discussions/17207#discussioncomment-81382

**Reply by rahul3103** [reply]
It needs to be on the same route. The route shouldn't change.

**Reply by leerob** [reply]
See above: https://github.com/vercel/next.js/discussions/17248#discussioncomment-82100

**Reply by andreas1327250** [reply]
You can't do it with native nextjs features. However, you can use a custom server (e. g. express) and route the requests to different pages based on the existance of an authentication-cookie.

**Comment by mrclay** [comment]
Would this not be a case for Preview Mode? getStaticProps() is dynamic if the preview mode cookie is present. Limitation is you can’t sniff the cookie client-side. Must return evidence of it from gSP(). Also performance is... unknown.
